### PR TITLE
Harden security

### PR DESCRIPTION
**Pull request by mindplay-dk**
The purpose of this WIP branch is to harden the security of `kodus/session`.

Work done so far:

 * added `Expires` to the Session Cookie (the minimum OWASP recommendation)
 * internally tracks whether `SessionData` instances are "new", as opposed to resumed by parsing an existing, valid Session Cookie.
 * no longer emits any Session Cookie when the Session is empty.
 * destroy resumed Sessions if empty, by revoking and expiring the Session Cookie.
 * add missing test + docs specifying that a Session ID is a valid UUID v4.
 * break down tests into smaller tests for each different scenario.

The goal is to implement [Renewal Timeout](https://www.owasp.org/index.php/Session_Management_Cheat_Sheet#Automatic_Session_Expiration) and comply better with OWASP recommendations.

The `SessionData` constructor already has a minor breaking change, so technically this should be a `3.0` release, even if it's highly unlikely anyone has been creating their own `SessionData` instances or extending the constructor, etc.

Next step is to harden the cookie against session fixation by adding an expiration timestamp, a salt, and a hash, to the cookie value: if somebody were to steal the cookie and learn the Session ID, they won't be able to access the Session once the cookie expires.

### CODE REVIEW

**Review by thomasnordahl-dk — COMMENTED** [review]
Adding expiration to cookie - good :)
Adding and breaking up tests - good :)

I'm not sure I feel totally awesome about the added isNew attribute to the sessionData class (simply because of added complexity), so I took a look a the reasons for adding it.
1. "no longer emits any Session Cookie when the Session is empty."  I'm not sure how this improves security or how it is otherwise harmful (not denying it, I just don't know any reason myself :-) ).
2. Only emitting the long lived cookie if the session is new, changes the behaviour somewhat, in that now a session only lives for ttl seconds, instead of ttl seconds after the last user activity. 

If any of those changes are crucial I don't have a better option than the one proposed here, but if we can do without  the changes mentioned above, I would prefer not adding the isNew attribute. Simply for simplicity ;-)

File: src/SessionService.php [hunk 1]
```diff
@@ -93,15 +94,46 @@ public function commitSession(SessionData $session, ResponseInterface $response)
...
 
-        $header = sprintf(self::COOKIE_NAME . "=%s; Path=/; HTTPOnly; SameSite=Lax;%s", $session_id, $secure);
+    private function createCookie(string $value, int $expires): string
+    {
+        $cookie = [
+            self::COOKIE_NAME . "={$value}",
+            "Path=/",
+            "HTTPOnly",
+            "SameSite=Lax",
+            "Expires=" . DateTime::createFromFormat("U", $expires, timezone_open('UTC'))->format(DateTime::COOKIE)
```
Comment (boan2010) [inline src/SessionService.php]: Do we consider this, now we are setting the time on the cookie: 

> Types of Cookies
There are two different types of cookies - session cookies and persistent cookies. If a cookie does not contain an expiration date, it is considered a session cookie. Session cookies are stored in memory and never written to disk. When the browser closes, the cookie is permanently lost from this point on. If the cookie contains an expiration date, it is considered a persistent cookie. On the date specified in the expiration, the cookie will be removed from the disk.

https://www.cisco.com/c/en/us/support/docs/security/web-security-appliance/117925-technote-csc-00.html#toc-hId-841855014

Comment (boan-jfm) [inline src/SessionService.php]: Do we consider this, now we are setting the time on the cookie: 

> Types of Cookies
There are two different types of cookies - session cookies and persistent cookies. If a cookie does not contain an expiration date, it is considered a session cookie. Session cookies are stored in memory and never written to disk. When the browser closes, the cookie is permanently lost from this point on. If the cookie contains an expiration date, it is considered a persistent cookie. On the date specified in the expiration, the cookie will be removed from the disk.

https://www.cisco.com/c/en/us/support/docs/security/web-security-appliance/117925-technote-csc-00.html#toc-hId-841855014

**Comment by mindplay-dk** [comment]
@boan2010 @thomasnordahl-dk thank you for your thorough review :-)

@thomasnordahl-dk

> "no longer emits any Session Cookie when the Session is empty." I'm not sure how this improves security or how it is otherwise harmful (not denying it, I just don't know any reason myself :-) ).

This change was made for a couple of reasons:

First, @boan2010 at some point noticed the Session ID changing with every request, which would happen with empty sessions - it looked like a bug to him, and arguably this behavior is a bit surprising; there's no practical reason to start empty sessions, if we can avoid it.

@rhl-jfm also recommended this change, mainly because there's no practical reason to issue a Session ID until there's actual session data.

In addition, this *might* help with certain GDPR legal issues, e.g. avoids tracking sessions before users actually interact with the site in some way, generating some data that needs to persist across requests.

> Only emitting the long lived cookie if the session is new, changes the behaviour somewhat, in that now a session only lives for ttl seconds, instead of ttl seconds after the last user activity

Yes, I'm still trying to figure out the appropriate renewal strategy.

But, from the [OWASP guide on session management](https://www.owasp.org/index.php/Session_Management_Cheat_Sheet#Automatic_Session_Expiration):

> All sessions should implement an absolute timeout, regardless of session activity. This timeout defines the maximum amount of time a session can be active, closing and invalidating the session upon the defined absolute period since the given session was initially created by the web application. After invalidating the session, the user is forced to (re)authenticate again in the web application and establish a new session.
> 
> The absolute session limits the amount of time an attacker can use a hijacked session and impersonate the victim user.

I don't know if our clients are going to accept or agree with that - they probably *don't* want users to exprerience *seemingly* random session terminations because we use a fixed timeout, but I don't know how to address that. Quietly issuing a renewed cookie would enable attackers to "fork" the session, e.g. obtaining a valid cookie from an existing stolen cookie, which they could then keep renewing forever, so I'm not sure how to deal with this.

I think we may have to shelve this change until @rhl-jfm returns from vacation and can weigh in.

**Pull request merged**